SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java

```diff
@@ -204,25 +170,65 @@ private void runLiDAR()
             {
                while(startRead)
                {
-            	   if(amountOfBytes() >= 9)
-            	   {
-            		   byte bytes[] = readBytes();
-            		   if(isValid(bytes))
-            		   {
-            			   distL = bytes[2];
-            			   distH = bytes[3];
-            			   strengthL = bytes[4];
-            			   strengthH = bytes[5];
-            			   reservedBytes = bytes[6];
-            			   signalQualityDegree = bytes[7];
-            		   }
-            	   }
+                   byte bytes[] = sensor.read(1);
+                   if(bytes.length > 0)
+                   {
+                       switch(state)
+                       {
+                           case WAIT_START1:
+                               if(bytes[0] == 0x59)
+                               {
+                                   count = 0;
+                                   checksum = bytes[0];
+                                   packet[0] = bytes[0];
+                                   state = ReceivingState.WAIT_START2;
+                               }
+                               break;
+                           case WAIT_START2:
+                               if(bytes[0] == 0x59)
+                               {
+                                   count++;
+                                   checksum += bytes[0];
+                                   packet[1] = bytes[0];
+                                   state = ReceivingState.RECEIVING;
+                               }
+                               else
+                               {
+                                   state = ReceivingState.WAIT_START1;
+                               }
+                               break;
+                           case RECEIVING:
+                               count++;
+                               packet[count] = bytes[0];
+                               if(count == PACKET_LENGTH - 1)
+                               {
+                                   // Reached the end of the packet, check checksum
+                                   if(checksum == bytes[0])
+                                   {
+                                       parsePacket();
+                                   }
+                                   state = ReceivingState.WAIT_START1;
+                               }
+                               else
+                               {
+                                   checksum += bytes[0];
+                               }
+                               break;
+                       }
+                   }
                }
             }
             catch(Exception e) {}
         }).start();
     }
     
+    public void parsePacket()
+    {
+        distance = (((int) packet[3] * 256) + ((int) packet[2]));
+        strength = (((int) packet[5] * 256) + ((int) packet[4]));
+        quality = (int)packet[7];
+    }
```

Comment:
Also, don't we need a synchronized(this) in here to make sure we're OK to write the new values and aren't switched out in the middle of these three writes?

Reply:
Probably a good idea - this should be a critical section (by whatever means you use to do this in Java). I'm not worried about new packets being read as we're processing old one since this can't happen given the way the state machine is structured but someone querying distance, strength or quality could get values from different packets in some cases (though this is unlikely to cause problems either, I suspect).